Homework 8 - Architecture patterns

fdemchenko opened the pull request:
Implement unsubscribe functionality.

Any advices are welcome, about moving send email trigger to mailer service! This part of home task causes difficulties, because I don't know how to organize data access properly.
Now, main service (with trigger functionality) sends message `ExchangeRateUpdated` to RabbitMQ with request to update email templates with new exchange rate. After that, service sends messages `SendEmailCommand` for each individual email to mailer service through message broker. How good this approach is?

![arch drawio (1)](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-fdemchenko/assets/89973537/07ac21b8-5bb9-4b95-96dd-58c11e3b23b6)

fdemchenko committed:
Add application architecture diagram.

fdemchenko committed:
Implement unsubscribe functionality

fdemchenko committed:
Fix queue declaration error, remove hardcoded value

fdemchenko committed:
Extract database setup logic into separate package

fdemchenko committed:
Create migrations package

fdemchenko committed:
Create basic customers service setup

MaxDMT16 commented:
> How good this approach is?

Although it's quite good approach, in task it's said to move trigger to email sender.
Could you implement this please?

fdemchenko commented:
Yeah, I'm working on it!

fdemchenko committed:
Create one more init sql script for user and privilliges setup

fdemchenko committed:
Fix AutoMigrate() function

fdemchenko committed:
Extract messaging logic into separate package. Create customers service related messages

fdemchenko committed:
Add customers service migrations

fdemchenko committed:
Add subscription ID to make compensate deletion after creaton of
customer is failed. Fix error with json tags

fdemchenko committed:
Change naming: emails -> subscriptions. Add DeleteByID() to subscriptions repository

fdemchenko committed:
Replace local openDB function with database package helper

fdemchenko committed:
Create generic RabbitMQ message publisher

fdemchenko committed:
Change naming email -> subscriptions

fdemchenko committed:
Add ability to return created subscription id

fdemchenko committed:
New customers service

fdemchenko committed:
Fix tests and naming

fdemchenko committed:
Start SAGA consumer and add publishing events on subscription to create customer account

fdemchenko committed:
Fix minor errors. Create Dockerfile for customers service

fdemchenko committed:
Add customers service to docker-compose.yml

fdemchenko committed:
Fix linter issues. Break down long lines

fdemchenko committed:
Fix invalid queue name error

fdemchenko committed:
Fix type

fdemchenko committed:
Move scheduler to mailer service

fdemchenko commented:
I made fast-forward merge with hw9-saga and this accidentally closed pull-request with SAGA implementation (because it was a base branch) and added unrelated commits here.
So, I've moved scheduler service and added unsubscribe functionality in this PR (commits: [a50d51d](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-fdemchenko/pull/9/commits/a50d51d1e71581926fcbd0e7ba6d6d29dba85664), [91c3dd1](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-fdemchenko/pull/9/commits/91c3dd1d58ca62c7475fca06c8adcb2b7d0a36ac))

fdemchenko commented:
Updated diagram:
![arch](https://github.com/user-attachments/assets/e102df37-96b0-4868-9f41-45fb24a35090)

fdemchenko committed:
Remove RabbitMQ connection creaton in every queue declare call

fdemchenko committed:
Add graceful shutdown to services

fdemchenko committed:
Remove cyclop linter

MaxDMT16 commented on cmd/mailer/internal/services/scheduler.go:
```diff
@@ -0,0 +1,53 @@
+}
+
+func (es *EmailScheduler) StartBackhroundTask(updateInterval time.Duration) {
+	ticker := time.NewTicker(updateInterval)
```
Most likely I've missed such implementation of scheduled action in previous PRs. However I've to notice it's not the most optimal way to implement. It's hard to control when messages will be sent cause it depends on start time of the mailer service. The only what is controlled with current implementation is frequency, but not absolute time of execution.

fdemchenko replied:
Yeah, agree. It's time to switch to CRON scheduled task

MaxDMT16 commented on cmd/mailer/Dockerfile:
```diff
@@ -24,4 +24,4 @@ WORKDIR /app
 
 # Set the entrypoint command
-ENTRYPOINT ["/app/mailer"]
+ENTRYPOINT ["/app/mailer", "--scheduler-interval", "15s"]
```
IMHO sending messages every 15 seconds is to frequent)

fdemchenko replied:
Forgot to remove after email testing, fixing it

MaxDMT16 reviewed (commented):
It's quite hard to review such big PRs that contains several home assignments (HA). For the future please try to make PR per one HA.
Also provided diagram seems to be outdated as it misses customers service. Despite on this additional comment to the diagram - all arrows point to message broker and it's not clear the flow of the messages in the system. Could you please update it and add to the project's readme, please?

fdemchenko committed:
Fix mailer update interval

fdemchenko commented:
I'm going to update app diagram and create CRON scheduler soon. Thanks for review!

fdemchenko committed:
Create CRON scheduler

fdemchenko committed:
Merge branch 'hw5-architecture' into hw8-architecture-patterns

fdemchenko committed:
Update arch diagram

fdemchenko committed:
Update dependencies

fdemchenko committed:
Add error value check for cron

fdemchenko committed:
Update README.md

fdemchenko committed:
Remove unused sql script

fdemchenko commented:
Done
